Duplicati/Library/Main/Database/LocalDatabase.cs

```diff
@@ -955,40 +960,41 @@ System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
         SELECT 
-            ""A"".""Path"" AS ""Path"", 
-            ""D"".""Lastmodified"" AS ""Lastmodified"", 
-            ""B"".""Length"" AS ""Filelength"", 
-            ""B"".""FullHash"" AS ""Filehash"", 
-            ""E"".""FullHash"" AS ""Metahash"", 
-            ""E"".""Length"" AS ""Metalength"", 
-            ""A"".""BlocksetID"" AS ""BlocksetID"",
-            ""F"".""Hash"" AS ""FirstBlockHash"",
-            ""F"".""Size"" AS ""FirstBlockSize"",
-            ""H"".""Hash"" AS ""FirstMetaBlockHash"",
-            ""H"".""Size"" AS ""FirstMetaBlockSize"",
-            ""C"".""BlocksetID"" AS ""MetablocksetID""
+	        ""A"".""Path"" AS ""Path"", 
+	        ""D"".""Lastmodified"" AS ""Lastmodified"", 
+	        ""B"".""Length"" AS ""Filelength"", 
+	        ""B"".""FullHash"" AS ""Filehash"", 
+	        ""E"".""FullHash"" AS ""Metahash"", 
+	        ""E"".""Length"" AS ""Metalength"",
+	        ""A"".""BlocksetID"" AS ""BlocksetID"",
+	        ""F"".""Hash"" AS ""FirstBlockHash"",
+	        ""F"".""Size"" AS ""FirstBlockSize"",
+	        ""H"".""Hash"" AS ""FirstMetaBlockHash"",
+	        ""H"".""Size"" AS ""FirstMetaBlockSize"",
+	        ""C"".""BlocksetID"" AS ""MetablocksetID""
         FROM 
-            ""File"" A, 
-            ""Blockset"" B, 
-            ""Metadataset"" C, 
-            ""FilesetEntry"" D, 
-            ""Blockset"" E, 
-            ""Block"" F,
-            ""BlocksetEntry"" G,
-            ""Block"" H,
-            ""BlocksetEntry"" I
+	        ""File"" A	
+        LEFT JOIN ""Blockset"" B
+          ON ""A"".""BlocksetID"" = ""B"".""ID"" 
+        LEFT JOIN ""Metadataset"" C  
+          ON ""A"".""MetadataID"" = ""C"".""ID""
+        LEFT JOIN ""FilesetEntry"" D
+          ON ""A"".""ID"" = ""D"".""FileID""
+        LEFT JOIN ""Blockset"" E
+          ON ""E"".""ID"" = ""C"".""BlocksetID""
+        LEFT JOIN ""BlocksetEntry"" G
+          ON ""B"".""ID"" = ""G"".""BlocksetID""
+        LEFT JOIN ""Block"" F 
+          ON ""G"".""BlockID"" = ""F"".""ID""  
+        LEFT JOIN ""BlocksetEntry"" I
+          ON ""E"".""ID"" = ""I"".""BlocksetID""
+        LEFT JOIN ""Block"" H 
+          ON ""I"".""BlockID"" = ""H"".""ID""
         WHERE 
-            ""A"".""ID"" = ""D"".""FileID"" 
-            AND ""D"".""FilesetID"" = ? 
-            AND ""A"".""BlocksetID"" = ""B"".""ID"" 
-            AND ""A"".""MetadataID"" = ""C"".""ID"" 
-            AND ""E"".""ID"" = ""C"".""BlocksetID""
-            AND ""B"".""ID"" = ""G"".""BlocksetID""
-            AND ""G"".""BlockID"" = ""F"".""ID""
-            AND ""G"".""Index"" = 0
-            AND ""I"".""BlocksetID"" = ""E"".""ID""
-            AND ""I"".""BlockID"" = ""H"".""ID""
-            AND ""I"".""Index"" = 0
+          ""A"".""BlocksetId"" >= 0 AND
+          ""D"".""FilesetID"" = ? AND
+          (""I"".""Index"" = 0 OR ""I"".""Index"" IS NULL) AND  
+          (""G"".""Index"" = 0 OR ""G"".""Index"" IS NULL)
         ) J
     LEFT OUTER JOIN 

```

Comment:
Frankly, I often find Duplicati SQL queries hard to follow :).

The rewrite is equivalent to the previous version with the following exceptions:
- Previously it used inner joins, now we're using left joins. This is needed because we might not have a block for all entries anymore. I could combine comma joins with explicit left joins but I find such queries even more cryptic.
 - Added a condition for non-negative blockset IDs. This is to filter out special blockset values (currently -100 and -200).
 - Added IS NULL conditions for possible indices. The purpose of the where clause is to always pick the first block, but what if we don't have any blocks -> allow null as well.